enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,476 @@
+
+- *Consumer*: A destination for logs, identified by a forwarder `output`.
+  The default consumer is the *Elasticsearch Log Store*
+- *Entry*: a single log entry, usually a single line of text. May be JSON or other format.
+- *Structured Log*: a log where each *entry* is formatted as a *JSON object*.
+- *Record*: A key-value record including the *entry* and meta-data collected by the logging system.
+  - *Internal record*: abstract record manipulated inside the forwarder.
+  - *Output record*: concrete record encoded by an `ouput` for a consumer. Encodings include JSON and Rsyslog.
+- *Schema*: Defines key names and value types in a structured entry.
+   The forwarder associates *schema names* with logs, it does not store or process user schema.
+- *Data Model*: The [schema][data_model] for the forwarder's own *record*.
+
+## Motivation
+
+### Goals
+
+* Direct access to JSON log entries as JSON objects in log records for `fluentdForward`, `Elasticsearch` or any other JSON-parsing consumer.
+* User can associate schema names with logs by category, namespace, application label, or any future forwarder `input` selection criteria.
+* Elasticsearch indexing of JSON logs with multiple schema without index explosions.
+* Upgrade path from today's implementation.
+
+### Non-Goals
+
+* General-purpose JSON queries and transformations.
+* Recording or validating user-defined schema.
+  The forwarder only identifies schema by name, all schema knowledge belongs to the user.
+
+## Proposal
+
+#### Data model
+
+Two new top-level fields added to the record [data model][data_model]:
+
+* `structured` (type object): Original structured JSON log entry.
+* `schema` (type string): Schema name assigned by user.
+
+Relationship between `structured` and `message` fields:
+
+* EXACTLY ONE of `message` or `structured` MUST be present.
+* `schema` is OPTIONAL, defined and interpreted by the user.
+* The presence or absence of the `structured` field can be used to detect structured/unstructured records.
+* If a filter (see below) removes the `structured` field, an equivalent `message` field is added by.
+
+With an *Elasticsearch* store, `schema` is copied to the existing `viaq_index_name` field to set the index.
+`viaq_index_name` will continue to be used for compatibility and upgrade, but will eventually be removed.
+
+#### ClusterLogForwarder configuration
+
+The elements of the CLF API are:
+
+- *Input*: Selects logs, adds meta-data, creates _internal_ records.
+- *Pipeline*: Transforms, filters and routes internal records.
+- *Output*: Connects to consumers, encodes and sends _output_ records.
+
+New fields:
+
+`input.stuctured`: (string, default "json") Attempt to parse structured entries.
+The only legal value is "json", others may be added in future.
+
+An input record contains:
+
+* `structured` field if `structured=json` *and* parse succeeds.
+* `message` field if `input.structured` is absent *or* JSON parse fails.
```

Comment:
in the case the JSON parse fails, do we expect this to end up in the same index as a similar record that didn't fail JSON parsing?

Reply:
Depends. The user can configure any of:
- drop unparsed records: pipeline {filter:{require:structured}}
- send unparsed records to a different output: second pipeline {filter{reject:structured}}
- send both parsed and unparsed to the same output: plain pipeline, no filter.
Probably the legacy ES output should tag them for different indexes (e.g. unstructured to app/infra/audit indices, structured to index-per-schema.
With ES pipelines we can move that descision to the store itself.

I need to talk to @lukas-vlcek  to finalize how we want to use index names.